File: sample-bots/python/MyFirstDroid/MyFirstDroid.py
from dataclasses import dataclass
from typing import Optional

from robocode_tank_royale.bot_api import Bot, team_message_type
from robocode_tank_royale.bot_api.color import Color
from robocode_tank_royale.bot_api.events import TeamMessageEvent
from robocode_tank_royale.bot_api.droid_abc import DroidABC


# ------------------------------------------------------------------
# Communication objects for team messages
# ------------------------------------------------------------------

# Point (x,y) class
@team_message_type
@dataclass
class Point:
    x: float
    y: float


# Robot colors
@team_message_type
@dataclass
class RobotColors:
    body_color: Optional[Color] = None
    tracks_color: Optional[Color] = None
    turret_color: Optional[Color] = None
    gun_color: Optional[Color] = None
    radar_color: Optional[Color] = None
    scan_color: Optional[Color] = None
    bullet_color: Optional[Color] = None


# ------------------------------------------------------------------
# MyFirstDroid
# ------------------------------------------------------------------
# A sample bot originally made for Robocode by Mathew Nelson.
#
# This is a droid bot meaning that it has more energy, but no radar.
# Member of the MyFirstTeam. Follows orders of team leader.
# ------------------------------------------------------------------
class MyFirstDroid(Bot, DroidABC):

    def run(self) -> None:
        print("MyFirstDroid ready")
        while self.running:
            # Execute next turn (on_team_message() handles the logic based on team messages)
            self.go()
        # terminates when this point is reached

    def on_team_message(self, e: TeamMessageEvent) -> None:
        message = e.message

        if isinstance(message, Point):
            # ------------------------------------------------------
            # Message is a point towards a target
            # ------------------------------------------------------

            # Turn body to target and fire hard
            self.turn_right(self.bearing_to(message.x, message.y))
            self.fire(3)

        elif isinstance(message, RobotColors):
            # ------------------------------------------------------
            # Message is containing new robot colors
            # ------------------------------------------------------

            # Read and set the robot colors
            self.body_color = message.body_color
            self.tracks_color = message.tracks_color
            self.turret_color = message.turret_color
            self.gun_color = message.gun_color
            self.radar_color = message.radar_color
            self.scan_color = message.scan_color
            self.bullet_color = message.bullet_color


def main() -> None:
    bot = MyFirstDroid()
    bot.start()


if __name__ == "__main__":
    main()

File: sample-bots/typescript/MyFirstDroid/MyFirstDroid.ts
import { Bot, ColorUtil, Droid, TeamMessageEvent } from "@robocode.dev/tank-royale-bot-api";

// ------------------------------------------------------------------
// MyFirstDroid
// ------------------------------------------------------------------
// A sample bot originally made for Robocode by Mathew Nelson.
//
// This is a droid bot meaning that it has more energy, but no radar.
// Member of the MyFirstTeam. Follows orders of team leader.
// ------------------------------------------------------------------

// Communication objects for team messages
interface BotPoint {
  type: "BotPoint";
  x: number;
  y: number;
}

interface BotColors {
  type: "BotColors";
  bodyColor: string;
  tracksColor: string;
  turretColor: string;
  gunColor: string;
  radarColor: string;
  scanColor: string;
  bulletColor: string;
}

type TeamMessage = BotPoint | BotColors;

class MyFirstDroid extends Bot implements Droid {
  static main() {
    new MyFirstDroid().start();
  }

  override run() {
    console.log("MyFirstDroid ready");
    while (this.isRunning()) {
      this.go(); // execute next turn (onTeamMessage() handles all bot logic based on team messages)
    }
  }

  // Called when a team message is received, which will be sent from MyFirstLeader
  override onTeamMessage(e: TeamMessageEvent) {
    const message = JSON.parse(e.message) as TeamMessage;
    if (message.type === "BotPoint") {
      // ------------------------------------------------------
      // Message is a point towards a target
      // ------------------------------------------------------
      this.turnRight(this.bearingTo(message.x, message.y));
      this.fire(3);
    } else if (message.type === "BotColors") {
      // ------------------------------------------------------
      // Message is containing new robot colors
      // ------------------------------------------------------
      this.setBodyColor(ColorUtil.fromHexColor(message.bodyColor));
      this.setTracksColor(ColorUtil.fromHexColor(message.tracksColor));
      this.setTurretColor(ColorUtil.fromHexColor(message.turretColor));
      this.setGunColor(ColorUtil.fromHexColor(message.gunColor));
      this.setRadarColor(ColorUtil.fromHexColor(message.radarColor));
      this.setScanColor(ColorUtil.fromHexColor(message.scanColor));
      this.setBulletColor(ColorUtil.fromHexColor(message.bulletColor));
    }
  }
}

MyFirstDroid.main();
